MAINT: Refactor clearing up of array data

seberg opened the pull request:
Replacement for gh-22692 and the "other option" here which does not reuse `ArrayMethod` as that seems a bit ridiculous for a single array operation.

This does a few things:

* Creates a new `get_clear_function()` slot, similar to the typical ArrayMethod strided loop getting, but a bit reduced.
   - Returns a simplified strided loop function (only one operand).
   - I have included a `NULL` initial argument, to have space for something like a "context"...
* Moves the current clearing functions that are based on the casting code out into its own file, this copies a little and moves some other things.
  - Recreates some helper infrastructure from the casting one...
* Introduces a new `PyArray_ClearBuffer` and `PyArray_ClearArray` and uses them.
  - *Effectively* replaces the old DECREF API, but I didn't dare delete it since it is in principle public.  We also use the single-item DECREF in a few places still. 
  - (The buffer version is almost unused now, after refactoring npyiter, but that seems OK)
* Fixes up NpyIter to track the `clear` function (may never be used, but it is actually nice to have it already in case of an error).

---

I suspect that this should be ready (modulo cleaning up the code a bit with review).  It does not yet expose this to user DTypes and a few code paths still use `PyArray_Item_XDECREF` which need fixing (I bet half of them are dubious anyway, but a different issue).

seberg committed:
MAINT: Refactor clear function to live on the DType

seberg committed:
MAINT: Reorganize array dealloc and nditer buffer cleanup

seberg committed:
WIP: Move traversal and simplify signature

seberg committed:
WIP: Further fixups and full implementation for structured dtypes

seberg committed:
Fixups, but npyiter is blocking me as it actually uses the "clear" logic once

seberg committed:
MAINT: Make buffer clearing explicitly stored in the iterator

This removes re-using `write` for the same purpose, which seems
surprisingly clean, considering that the "clearing" path was
already special in any case.

seberg commented:
@ngoldbaum I don't know if you dare review this, but if you do that would be great and we could also go through together or so.  I think the main point is deciding that final API is fine, the rest is tricky code, but should be OK.

seberg committed:
BUG: Readd accidentally remove function return type

Also remove accidentally auto-inserted include...

ngoldbaum commented on numpy/core/src/multiarray/array_method.c:
```diff
@@ -334,28 +334,29 @@ fill_arraymethod_from_slots(
         PyErr_Format(PyExc_TypeError,
-                "Must provide a strided inner loop function. (method: %s)",
+                "Must provide unaligned strided inner loop for method to "
+                "indicate unaligned support (method: %s)",
```
It might be a bit clearer to dtype authors if you had a different error message for the case when someone defines an `unaligned_strided_loop` but doesn't set the flag correctly.

ngoldbaum commented on numpy/core/src/multiarray/dtype_traversal.c:
```diff
@@ -0,0 +1,452 @@
+ * Python object DECREF/dealloc followed by NULL'ing the data
+ * (to support double clearing on errors).
+ * However, memory initialization and traverse follows similar
+ * protocols (although traversal needs additional arguments.
```
missing close paren

ngoldbaum commented on numpy/core/src/multiarray/dtype_traversal.h:
```diff
@@ -0,0 +1,105 @@
+
+/*
+ * A simplified loop, similar to a general strided-loop function.
+ * Using a `void *reserved`, because I think we probably need to pass in
```
Do you mean `void *traverse_context` here?

ngoldbaum commented on numpy/core/src/multiarray/dtypemeta.h:
```diff
@@ -58,7 +61,13 @@ typedef struct {
+    PyArrayMethodObject *within_dtype_castingimpl;
+    /*
+     * Implementation which clears a dtype or NULL.  If not given, setting
+     * HASREF on the dtype is invalid.  We only use the loop getting, not
```
As far as I can see, `HASREF` isn't a thing, so this comment might be a bit confusing. It looks like `PyDataType_REFCHK` checks `NPY_ITEM_REFCOUNT`, so you could use that? Although maybe `NPY_ITEM_IS_POINTER` is invalid too?

seberg replied:
Yeah, `NPY_ITEM_REFCOUNT` is it, except that we may want to rename/alias it to `NPY_ITEM_HASREF` or something...

To be honest, I am not sure that `NPY_ITEM_IS_POINTER` is ever used (or useful) except  as having the identical meaning to `NPY_ITEM_REFCOUNT`...

ngoldbaum replied:
Ah that makes sense. I brought up `NPY_ITEM_IS_POINTER` because it might be nice to indicate that a dtype uses embedded references but doesn't need the GIL.

ngoldbaum commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -1153,26 +1135,7 @@ _new_argsortlike(PyArrayObject *op, int axis, PyArray_ArgSortFunc *argsort,
-                copyswapn(valbuffer, elsize,
-                          it->dataptr, astride, N, swap, op);
-            }
+            copyswapn(valbuffer, elsize, it->dataptr, astride, N, swap, op);
```
nice cleanup

seberg replied:
Oh, I didn't remember this sneaking into this PR, but I guess it is just as well...

ngoldbaum commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -21,6 +25,94 @@ static void
+        PyArray_Descr *descr, char *data,
+        npy_intp stride, npy_intp size, int aligned)
+{
+    if (!PyDataType_REFCHK(descr)) {
```
Should this also check `NPY_ITEM_IS_POINTER`?

ngoldbaum commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -204,6 +296,10 @@ PyArray_INCREF(PyArrayObject *mp)
+
+  The use of this function is strongly discouraged, within NumPy
+  use PyArray_Clear, which DECREF's and sets everything to NULL and can
+  work with any dtype.
```
Should (can?) `PyArray_XDECREF` be deprecated?

seberg replied:
Probably, not sure there is a need for it to begin in reasonable code... But seems hard to do, might be a "nuke it with NumPy 2.0" thing...

ngoldbaum commented on numpy/core/src/multiarray/dtype_traversal.h:
```diff
@@ -0,0 +1,105 @@
+ * We assume for now that this context can be just passed through in the
+ * the future (for structured dtypes).
+ */
+typedef int (simple_loop_function)(
```
nit: maybe these two new function signatures should be called `traverse_loop_function` and `get_traverse_loop_function` to indicate that it's a loop specifically for array traversal

seberg replied:
Fair point, "simple" is always a bad name anyway ;).

ngoldbaum reviewed (commented):
The new API looks fine to me, I think. I can't think of a reason why you'd want to use the auxdata for clearing the array, but it doesn't seem harmful to allow it to be passed in if someone wants it.

Currently I think this only works with embedded references to python objects, but in principle it should work with arbitrary embedded references. Does that just mean expanding the checks to include `NPY_ITEM_IS_POINTER` or are there any other possible issues?

Presumably user dtypes will also need a way to set the appropriate flags to indicate that they have embedded references?

seberg committed:
MAINT: Address review comments

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

seberg commented:
Updated based on the comments, thanks!  `auxdata` is already used for structured dtypes (we need to store how to unpack the struct).

It may also be useful for other things, for example sorting could re-use this pattern (at least partially).  Not quite a "traversal" I guess, but needs much the same.

seberg commented:
Gentle ping on this clearing of array data PR.

mattip commented on numpy/core/src/multiarray/refcount.c:
```diff
@@ -21,6 +25,94 @@ static void
+
+    int res = clear_info.func(
+            NULL, clear_info.descr, data, size, stride, clear_info.auxdata);
+    NPY_traverse_info_xfree(&clear_info);
```
What are the flags for?

seberg replied:
Future proofing and symmetry.  The only flags we have are requiring GIL and floating point errors.  Neither are relevant for clearing I think (I guess releasing GIL could be in principle, so maybe I could add a check for that).

I hope that the pattern could be used e.g. for sorting where flags would be more meaningful.

mattip reviewed (commented):
This could use some documentation about how users are meant to design "clear" functions, how "clear" relates to "dealloc" and "decref". Or is there documentation in one of the dtype NEPS?

seberg commented:
> This could use some documentation about how users are meant to design "clear" functions, how "clear" relates to "dealloc" and "decref". Or is there documentation in one of the dtype NEPS?

True, for users I didn't add any public API yet though, so I was thinking to add that together with that.  OTOH, if we want internal docs than should definitely add it here!

seberg committed:
DOC: Add code comment about why flags are not used in refcount.c

We simply don't bother releasing the GIL right now.  This could make
sense in principle for some user dtypes, though.
(NumPy only ever uses clearing with Python objects so the GIL is always
needed).

I see no plausible use of checking floating point errors, so do not
bother to add such cruft, if anyone ever needs it, they should add it.

seberg committed:
TST: Do not use `dtypemeta.h` in tests, they should use "external" API

mattip commented:
> if we want internal docs

I would at least like a type of glossary. How does "clearing" relate to "dealloc" or "decref"?

seberg committed:
DOC: Expand comment for `get_clear_loop` and sprinkle dealloc+nulling

seberg commented:
I added it to the `get_clear_loop` and mention `decref+NULLing` in a few more places.  (except the last commit, this is just a rebase).

seberg commented:
> Do we have a way to tell coverage "ignore this"?

I am not aware, although I suppose there probably is a way.  But coverage is only reported as changed so it never bothered me?

mattip commented on numpy/core/src/multiarray/dtype_traversal.c:
```diff
@@ -0,0 +1,455 @@
+
+/* transfer data copy function */
+static NpyAuxData *
+fields_clear_data_clone(NpyAuxData *data)
```
There is no test of this functionality, maybe it should be marked `/* untested */`. Or is it tested in the userdtype repo?

seberg replied:
This is weird internal API that is hard to use.  Let me try, it may actually be possible via `nditer`...

mattip commented on numpy/core/src/multiarray/dtype_traversal.c:
```diff
@@ -0,0 +1,455 @@
+    npy_intp i, field_count = d->field_count;
+
+    /* Do the traversing a block at a time for better memory caching */
+    const npy_intp blocksize = NPY_LOWLEVEL_BUFFER_BLOCKSIZE;
```
Are you sure this is more performant than a simpler approach?

seberg replied:
This is the approach we use everywhere.  But yes, it should make sense for cache friendliness.

mattip commented on numpy/core/src/multiarray/dtype_traversal.c:
```diff
@@ -0,0 +1,455 @@
+
+/* transfer data copy function */
+static NpyAuxData *
+subarray_clear_data_clone(NpyAuxData *data)
```
This is never used?

seberg replied:
Yes, this one actually seems unreachable, added a copy.  I think I covered most of the others (but its complicated, will have to see what is actually covered).

mattip commented on numpy/core/src/multiarray/dtypemeta.h:
```diff
@@ -58,7 +61,20 @@ typedef struct {
+     * If the DType class does not implement `get_clear_loop` setting
+     * NPY_ITEM_REFCOUNT on its dtype instances is invalid.  Note that it is
+     * acceptable for  NPY_ITEM_REFCOUNT to inidicate references that are not
+     * Python objects.
```
I think we should change this to make `NPY_ITEM_HASOBJECT` the preferred alias, and replace `PyDataType_REFCHK` with `PyDataType_HASOBJECT`, but that can be done in a future PR since this is all internal API.

mattip reviewed (commented):
Mainly concerned about the lack of coverage, it seems the copy functions are never used? Other than that this looks good.

seberg commented on numpy/core/src/multiarray/dtype_traversal.c:
```diff
@@ -0,0 +1,455 @@
+
+
+/* Same as in dtype_transfer.c */
+#define NPY_LOWLEVEL_BUFFER_BLOCKSIZE  128
```
It is the identical use case, but doesn't need to be the same value.

charris commented:
close/reopen

seberg commented on numpy/core/src/multiarray/dtype_transfer.c:
```diff
@@ -228,8 +217,8 @@ _strided_to_strided_any_to_object(
-        if (data->decref_src.func(&data->decref_src.context,
-                &orig_src, &N, &src_stride, data->decref_src.auxdata) < 0) {
+        if (data->decref_src.func(NULL, data->decref_src.descr,
+                orig_src, N, src_stride, data->decref_src.auxdata) < 0) {
```
Hmmm, it might be possible to hit this for structured to object (I suspect not, but I will just try to add a test).  And we only have object and structured, user dtypes could (in the future) hit it, though.

seberg committed:
Apply suggestions from code review

Co-authored-by: Matti Picus <matti.picus@gmail.com>

seberg committed:
TST: Add test and comment about path that seems uncoverable

seberg committed:
DOC: Modify comments about untested functions

We only ever use the "clone" mechanism for `nditer.copy()` but even
there, it is only used for buffer transfers (not final cleanups).

It appears that structured dtypes are always a single cast step
which doesn't require the clearing normally (so this is hard to
hit).

seberg committed:
TST: Add some more test coverage (but probably not the big clone funcs)

seberg committed:
MAINT: "comment out" unreachable subarray clear data clone

mattip commented:
Let's give this a try. It might take a few iterations to find all the corner cases, maybe the userdtype repo will find some of them.

mattip commented:
Thanks @seberg

seberg commented:
Thanks!  @ngoldbaum do you want to look into making this public?  (just fishing, but it should be very easy.)

ngoldbaum commented:
> Thanks! @ngoldbaum do you want to look into making this public? (just fishing, but it should be very easy.)

Sure, I can try that out. I'll ping you if I run into trouble.

mhvk commented:
@seberg - This PR is causing failures failures over at astropy - see https://github.com/astropy/astropy/pull/14460 (and maybe https://github.com/pandas-dev/pandas/issues/51589 as well, given link above; cc @lithomas1)

lithomas1 commented:
I couldn't reproduce on my computer, but the backtrace from our tests is is
```
0x00007ffff588c71b in OBJECT_copyswapn () from /home/runner/micromamba/envs/test/lib/python3.10/site-packages/numpy/core/_multiarray_umath.cpython-310-x86_64-linux-gnu.so
#0  0x00007ffff588c71b in OBJECT_copyswapn () from /home/runner/micromamba/envs/test/lib/python3.10/site-packages/numpy/core/_multiarray_umath.cpython-310-x86_64-linux-gnu.so
#1  0x00007ffff5915c72 in _new_argsortlike () from /home/runner/micromamba/envs/test/lib/python3.10/site-packages/numpy/core/_multiarray_umath.cpython-310-x86_64-linux-gnu.so
#2  0x00007ffff591a16d in PyArray_ArgSort () from /home/runner/micromamba/envs/test/lib/python3.10/site-packages/numpy/core/_multiarray_umath.cpython-310-x86_64-linux-gnu.so
#3  0x00007ffff59690bf in array_argsort () from /home/runner/micromamba/envs/test/lib/python3.10/site-packages/numpy/core/_multiarray_umath.cpython-310-x86_64-linux-gnu.so
#4  0x000055555568bd[83](https://github.com/pandas-dev/pandas/actions/runs/4262938751/jobs/7419045241#step:8:85) in cfunction_vectorcall_FASTCALL_KEYWORDS (func=func@entry=<built-in method argsort of numpy.ndarray object at remote 0x7fffb456e550>, args=0x55555ff63278, nargsf=<optimized out>, kwnames=('kind',)) at /usr/local/src/conda/python-3.10.9/Objects/methodobject.c:446
```

Maybe this helps a little? 

I only found this PR from git blame, where it seems like the only function to touch the functions in the backtrace recently.

mhvk commented:
In our case, the error message "RuntimeError: Internal error, tried to fetch clear function for the user dtype '�' without fields or subarray (legacy support)" is triggered, which was introduced here... I'm trying to see if this dtype actually has objects inside or not (which is what this PR addresses), but am not  very familiar with this part of astropy. Anyway, maybe @seberg has an idea just given what path seems to be hit...

mhvk commented:
p.s. The test that triggers the problem may well have an empty structured array, since it explicitly reads in a table without columns. Given an installation of astropy, a minimal example to reproduce is at https://github.com/astropy/astropy/pull/14460#issuecomment-1444936926

mhvk commented:
OK, found a minimal pure numpy example - see #23277

seberg commented:
@lithomas1 thanks.  Seems like I forgot to add a NULLing of the buffer when trying to simplify that code.  Odd that NumPy test suite doesn't notice that.

seberg commented:
Triggered the wheel build, so this hopefully should clear up very soon.  Thanks for the reports.

mhvk commented:
Thanks for the quick fix!
